### PR TITLE
Add GLO ephemeris

### PR DESCRIPTION
**Pull request by rge-exafore**
This commit adds definition of Glonass ephemeris data structure.

### CODE REVIEW

**Comment by ljbade** [comment]
@rge-exafore I would base the GLONASS changes on https://github.com/swift-nav/libswiftnav/pull/269 which should hopefully get merged soon. It tidies up the ephemeris and almanac functions and structs.

File: include/libswiftnav/ephemeris.h [hunk 1]
```diff
@@ -18,6 +18,9 @@
 #include <libswiftnav/time.h>
 #include <libswiftnav/common.h>
 
+#define SV_TYPE_GLONASS    0
+#define SV_TYPE_GLONASS_M  1
```
Comment (ljbade) [inline include/libswiftnav/ephemeris.h]: We currently don't capture this for GPS (but there is a PR for it in the L2C stuff). Should we store it the same way?

File: include/libswiftnav/ephemeris.h [hunk 2]
```diff
@@ -39,6 +42,34 @@ typedef struct {
...
 typedef struct {
+  /*u8 m;*/     /* string number */
+  /*u8 B;*/     /* unhealthy SV indication, should be stored in ephemeris_t */
+  /*u8 n;*/     /* SV id , should be stored in ephemeris_t*/
+  /*u8 l;*/     /* unhealthy SV indication, should be stored in ephemeris_t */
+  /*u8 P;*/     /* SV operation mode in respect of time parameters TODO: decide is we need to store it */
+
+  u8 tk_hours;            /* Current frame start time since beginning of day */
+  u8 tk_minutes;          /* Not needed in ephemeris. TODO: decide if it needs to be stored here. */
+  u8 tk_half_minutes;
```
Comment (ljbade) [inline include/libswiftnav/ephemeris.h]: I would store these in `toe`

File: include/libswiftnav/ephemeris.h [hunk 3]
```diff
@@ -39,6 +42,34 @@ typedef struct {
...
+
+  u16 tb;                 /* Index of the time interval within current day */
+  u8 sv_type;             /* Coded as word M in broadcasted ephemeris */
+  double gamma;
+  double tau;
+  double pos[3];          /* Position of the SV at tb in PZ-90.02 coordinates system, km */
+  double vel[3];          /* Velocity vector of the SV at tb in PZ-90.02 coordinates system, km/s */
+  double acc[3];          /* Acceleration vector of the SV at tb in PZ-90.02 coordinates system, km/s/s */
+  u16 NT;                 /* Current date, coded */
+  u16 accuracy;           /* User range accuracy coded in Ft word, meters */
```
Comment (ljbade) [inline include/libswiftnav/ephemeris.h]: use `ura` in base struct

File: include/libswiftnav/ephemeris.h [hunk 4]
```diff
@@ -39,6 +42,34 @@ typedef struct {
...
+  double delta_tau;       /* Time difference between L1 and L2 signals */
+  u8 age;                 /* Age of the information. Time since calculation till tb in days. */
```
Comment (ljbade) [inline include/libswiftnav/ephemeris.h]: We don't currently capture this for GPS, but might be useful in the future. If we add for GLONASS we should go back and also add for GPS

File: include/libswiftnav/ephemeris.h [hunk 5]
```diff
@@ -39,6 +42,34 @@ typedef struct {
...
+  u8 tk_half_minutes;
+
+  u16 tb;                 /* Index of the time interval within current day */
+  u8 sv_type;             /* Coded as word M in broadcasted ephemeris */
+  double gamma;
+  double tau;
+  double pos[3];          /* Position of the SV at tb in PZ-90.02 coordinates system, km */
+  double vel[3];          /* Velocity vector of the SV at tb in PZ-90.02 coordinates system, km/s */
+  double acc[3];          /* Acceleration vector of the SV at tb in PZ-90.02 coordinates system, km/s/s */
+  u16 NT;                 /* Current date, coded */
```
Comment (ljbade) [inline include/libswiftnav/ephemeris.h]: Is this part of `toe`?

File: include/libswiftnav/ephemeris.h [hunk 6]
```diff
@@ -39,6 +42,34 @@ typedef struct {
...
+  /*u8 P;*/     /* SV operation mode in respect of time parameters TODO: decide is we need to store it */
+
+  u8 tk_hours;            /* Current frame start time since beginning of day */
+  u8 tk_minutes;          /* Not needed in ephemeris. TODO: decide if it needs to be stored here. */
+  u8 tk_half_minutes;
+
+  u16 tb;                 /* Index of the time interval within current day */
+  u8 sv_type;             /* Coded as word M in broadcasted ephemeris */
+  double gamma;
+  double tau;
```
Comment (ljbade) [inline include/libswiftnav/ephemeris.h]: what is `gamma` and `tau`

File: include/libswiftnav/ephemeris.h [hunk 7]
```diff
@@ -39,6 +42,34 @@ typedef struct {
...
+  double pos[3];          /* Position of the SV at tb in PZ-90.02 coordinates system, km */
```
Comment (ljbade) [inline include/libswiftnav/ephemeris.h]: `pos` vel` `acc` could use the xyz struct

File: include/libswiftnav/ephemeris.h [hunk 8]
```diff
@@ -39,6 +42,34 @@ typedef struct {
...
+  double vel[3];          /* Velocity vector of the SV at tb in PZ-90.02 coordinates system, km/s */
+  double acc[3];          /* Acceleration vector of the SV at tb in PZ-90.02 coordinates system, km/s/s */
+  u16 NT;                 /* Current date, coded */
+  u16 accuracy;           /* User range accuracy coded in Ft word, meters */
+  double delta_tau;       /* Time difference between L1 and L2 signals */
```
Comment (ljbade) [inline include/libswiftnav/ephemeris.h]: We don't currently capture this for GPS, but will need for L2C. We should move this to the base struct and fill out for GPS too. (GPS calls it `tgd`)

File: include/libswiftnav/ephemeris.h [hunk 9]
```diff
@@ -39,6 +42,34 @@ typedef struct {
...
+  u8 age;                 /* Age of the information. Time since calculation till tb in days. */
+  u8 P1;                  /* Size of a time interval for index tb, minutes */
+  u8 P2;                  /* Oddness ("1") or evenness ("0") of the value of tb for intervals 30 or 60 minutes */
```
Comment (ljbade) [inline include/libswiftnav/ephemeris.h]: is this used to calculate `toe`?

File: include/libswiftnav/ephemeris.h [hunk 10]
```diff
@@ -39,6 +42,34 @@ typedef struct {
...
+  u8 P3;                  /* A number of almanacs in given frame. "1" - for 5 SVs, "0" - for 4 SVs */
```
Comment (ljbade) [inline include/libswiftnav/ephemeris.h]: not needed here, only in the decode function for almanac (which we don't even have for GPS yet)

File: include/libswiftnav/ephemeris.h [hunk 11]
```diff
@@ -39,6 +42,34 @@ typedef struct {
...
+  u8 P4;                  /* Flag indicating that a new ephemeris is present */
```
Comment (ljbade) [inline include/libswiftnav/ephemeris.h]: We don't capture this for GPS, but we could. Will be useful along with `age` in the future. If we add, should put in the base struct

**Comment by ljbade** [comment]
@rge-exafore I added some comments

File: include/libswiftnav/ephemeris.h (hunk 1 above)
Comment (rge-exafore) [inline include/libswiftnav/ephemeris.h]: Do you mean L2C capabilities PR? I don't actually know what is information on GLO SV type is for. So, I'll drop it from the ephemeris. If it indicates G2 capability we'll add it the same way we did for L2C capability in GPS.

File: include/libswiftnav/ephemeris.h (hunk 4 above)
Comment (rge-exafore) [inline include/libswiftnav/ephemeris.h]: I'll add it to a list of issues and drop for the time being from GLO ephe.

File: include/libswiftnav/ephemeris.h (hunk 5 above)
Comment (rge-exafore) [inline include/libswiftnav/ephemeris.h]: Yes, move to toe.

File: include/libswiftnav/ephemeris.h (hunk 6 above)
Comment (rge-exafore) [inline include/libswiftnav/ephemeris.h]: Added description

File: include/libswiftnav/ephemeris.h (hunk 7 above)
Comment (rge-exafore) [inline include/libswiftnav/ephemeris.h]: Didn't quite get that. Do you mean using ephemeris_xyz_t in ephemeris_glo_t? Or one more struct to be used in ephemeris_xyz_t and in ephemeris_glo_t? I don't like either of those as the first adds unnecessary fields to GLO ephe and the second requires quite a bit of changes in the code, in all places accessing those fields. So, I'll leave it as is for now, let's discuss. I change units to meters, though.

File: include/libswiftnav/ephemeris.h (hunk 8 above)
Comment (rge-exafore) [inline include/libswiftnav/ephemeris.h]: Added tgd to ephemeris_t.

File: include/libswiftnav/ephemeris.h (hunk 9 above)
Comment (rge-exafore) [inline include/libswiftnav/ephemeris.h]: Yes, removed.

File: include/libswiftnav/ephemeris.h (hunk 10 above)
Comment (rge-exafore) [inline include/libswiftnav/ephemeris.h]: ok, removed.

File: include/libswiftnav/ephemeris.h (hunk 11 above)
Comment (rge-exafore) [inline include/libswiftnav/ephemeris.h]: Will mention that in the issue recorded for the 'age' field.

**Comment by ljbade** [comment]
Looks good. However now that `ephemeris_xyz_t` and `ephemeris_glo_t` are basically identical we should just use xyz for GLONASS for the struct.

But in the `calc_sat_state` function we will call a different function for SBAS or GLONASS.

**Comment by rge-exafore** [comment]
@ljbade : we can do what you suggested, of course, but what's the reason to use weird field names in GLO ephemeris instead of adding proper structure definition?

**Comment by ljbade** [comment]
Looks good to me, I'll ask @jacobmcnamee to take a quick look too

**Comment by jacobmcnamee** [comment]
lgtm

**Pull request merged**